<!DOCTYPE html><html lang="zh-CN" data-theme="light"><head><meta charset="UTF-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no"><title>FACETS Testing CNV | Solivehong Blog</title><meta name="keywords" content="生信"><meta name="author" content="Solivehong"><meta name="copyright" content="Solivehong"><meta name="format-detection" content="telephone=no"><meta name="theme-color" content="ffffff"><meta name="description" content="FACETSfacets is CNV detection software with using the ascn principle  inputfirst Let’s look at the test data given   123456Chromosome Chromosome of the SNPPosition Position of the SNPFile1R Read depth">
<meta property="og:type" content="article">
<meta property="og:title" content="FACETS Testing CNV">
<meta property="og:url" content="https://zhaohongqiangsoliva.github.io/2020/06/09/2020-06-09-CNV-Facets%E8%BD%AF%E4%BB%B6%E6%A3%80%E6%B5%8Bcnv/index.html">
<meta property="og:site_name" content="Solivehong Blog">
<meta property="og:description" content="FACETSfacets is CNV detection software with using the ascn principle  inputfirst Let’s look at the test data given   123456Chromosome Chromosome of the SNPPosition Position of the SNPFile1R Read depth">
<meta property="og:locale" content="zh_CN">
<meta property="og:image" content="https://zhaohongqiangsoliva.github.io/img/dna.png">
<meta property="article:published_time" content="2020-06-08T16:00:00.000Z">
<meta property="article:modified_time" content="2022-09-06T10:11:41.322Z">
<meta property="article:author" content="Solivehong">
<meta property="article:tag" content="生信">
<meta name="twitter:card" content="summary">
<meta name="twitter:image" content="https://zhaohongqiangsoliva.github.io/img/dna.png"><link rel="shortcut icon" href="/img/favicon.png"><link rel="canonical" href="https://zhaohongqiangsoliva.github.io/2020/06/09/2020-06-09-CNV-Facets%E8%BD%AF%E4%BB%B6%E6%A3%80%E6%B5%8Bcnv/"><link rel="preconnect" href="//cdn.jsdelivr.net"/><link rel="preconnect" href="//busuanzi.ibruce.info"/><link rel="stylesheet" href="/css/index.css"><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/@fortawesome/fontawesome-free/css/all.min.css" media="print" onload="this.media='all'"><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/@fancyapps/ui/dist/fancybox.min.css" media="print" onload="this.media='all'"><script>const GLOBAL_CONFIG = { 
  root: '/',
  algolia: undefined,
  localSearch: {"path":"/search.xml","preload":true,"languages":{"hits_empty":"找不到您查询的内容：${query}"}},
  translate: undefined,
  noticeOutdate: undefined,
  highlight: {"plugin":"highlighjs","highlightCopy":true,"highlightLang":true,"highlightHeightLimit":false},
  copy: {
    success: '复制成功',
    error: '复制错误',
    noSupport: '浏览器不支持'
  },
  relativeDate: {
    homepage: false,
    post: false
  },
  runtime: '',
  date_suffix: {
    just: '刚刚',
    min: '分钟前',
    hour: '小时前',
    day: '天前',
    month: '个月前'
  },
  copyright: undefined,
  lightbox: 'fancybox',
  Snackbar: undefined,
  source: {
    justifiedGallery: {
      js: 'https://cdn.jsdelivr.net/npm/flickr-justified-gallery/dist/fjGallery.min.js',
      css: 'https://cdn.jsdelivr.net/npm/flickr-justified-gallery/dist/fjGallery.min.css'
    }
  },
  isPhotoFigcaption: false,
  islazyload: false,
  isAnchor: false
}</script><script id="config-diff">var GLOBAL_CONFIG_SITE = {
  title: 'FACETS Testing CNV',
  isPost: true,
  isHome: false,
  isHighlightShrink: false,
  isToc: true,
  postUpdate: '2022-09-06 18:11:41'
}</script><noscript><style type="text/css">
  #nav {
    opacity: 1
  }
  .justified-gallery img {
    opacity: 1
  }

  #recent-posts time,
  #post-meta time {
    display: inline !important
  }
</style></noscript><script>(win=>{
    win.saveToLocal = {
      set: function setWithExpiry(key, value, ttl) {
        if (ttl === 0) return
        const now = new Date()
        const expiryDay = ttl * 86400000
        const item = {
          value: value,
          expiry: now.getTime() + expiryDay,
        }
        localStorage.setItem(key, JSON.stringify(item))
      },

      get: function getWithExpiry(key) {
        const itemStr = localStorage.getItem(key)

        if (!itemStr) {
          return undefined
        }
        const item = JSON.parse(itemStr)
        const now = new Date()

        if (now.getTime() > item.expiry) {
          localStorage.removeItem(key)
          return undefined
        }
        return item.value
      }
    }
  
    win.getScript = url => new Promise((resolve, reject) => {
      const script = document.createElement('script')
      script.src = url
      script.async = true
      script.onerror = reject
      script.onload = script.onreadystatechange = function() {
        const loadState = this.readyState
        if (loadState && loadState !== 'loaded' && loadState !== 'complete') return
        script.onload = script.onreadystatechange = null
        resolve()
      }
      document.head.appendChild(script)
    })
  
      win.activateDarkMode = function () {
        document.documentElement.setAttribute('data-theme', 'dark')
        if (document.querySelector('meta[name="theme-color"]') !== null) {
          document.querySelector('meta[name="theme-color"]').setAttribute('content', '#0d0d0d')
        }
      }
      win.activateLightMode = function () {
        document.documentElement.setAttribute('data-theme', 'light')
        if (document.querySelector('meta[name="theme-color"]') !== null) {
          document.querySelector('meta[name="theme-color"]').setAttribute('content', 'ffffff')
        }
      }
      const t = saveToLocal.get('theme')
    
          if (t === 'dark') activateDarkMode()
          else if (t === 'light') activateLightMode()
        
      const asideStatus = saveToLocal.get('aside-status')
      if (asideStatus !== undefined) {
        if (asideStatus === 'hide') {
          document.documentElement.classList.add('hide-aside')
        } else {
          document.documentElement.classList.remove('hide-aside')
        }
      }
    
    const detectApple = () => {
      if(/iPad|iPhone|iPod|Macintosh/.test(navigator.userAgent)){
        document.documentElement.classList.add('apple')
      }
    }
    detectApple()
    })(window)</script><meta name="generator" content="Hexo 6.2.0"></head><body><div id="web_bg"></div><div id="sidebar"><div id="menu-mask"></div><div id="sidebar-menus"><div class="avatar-img is-center"><img src="/img/dna.png" onerror="onerror=null;src='/img/friend_404.gif'" alt="avatar"/></div><div class="sidebar-site-data site-data is-center"><a href="/archives/"><div class="headline">文章</div><div class="length-num">19</div></a><a href="/tags/"><div class="headline">标签</div><div class="length-num">4</div></a><a href="/categories/"><div class="headline">分类</div><div class="length-num">0</div></a></div><hr/><div class="menus_items"><div class="menus_item"><a class="site-page" href="/"><i class="fa-fw fas fa-home"></i><span> Home</span></a></div><div class="menus_item"><a class="site-page" href="/archives/"><i class="fa-fw fas fa-archive"></i><span> Archives</span></a></div><div class="menus_item"><a class="site-page" href="/tags/"><i class="fa-fw fas fa-tags"></i><span> Tags</span></a></div><div class="menus_item"><a class="site-page" href="/categories/"><i class="fa-fw fas fa-folder-open"></i><span> Categories</span></a></div><div class="menus_item"><a class="site-page group" href="javascript:void(0);"><i class="fa-fw fas fa-list"></i><span> List</span><i class="fas fa-chevron-down"></i></a><ul class="menus_item_child"><li><a class="site-page child" href="/gallery/"><i class="fa-fw fas fa-images"></i><span> Gallery</span></a></li></ul></div><div class="menus_item"><a class="site-page" href="/link/"><i class="fa-fw fas fa-link"></i><span> Link</span></a></div><div class="menus_item"><a class="site-page" href="/about/"><i class="fa-fw fas fa-heart"></i><span> About</span></a></div></div></div></div><div class="post" id="body-wrap"><header class="post-bg" id="page-header"><nav id="nav"><span id="blog_name"><a id="site-name" href="/">Solivehong Blog</a></span><div id="menus"><div id="search-button"><a class="site-page social-icon search"><i class="fas fa-search fa-fw"></i><span> 搜索</span></a></div><div class="menus_items"><div class="menus_item"><a class="site-page" href="/"><i class="fa-fw fas fa-home"></i><span> Home</span></a></div><div class="menus_item"><a class="site-page" href="/archives/"><i class="fa-fw fas fa-archive"></i><span> Archives</span></a></div><div class="menus_item"><a class="site-page" href="/tags/"><i class="fa-fw fas fa-tags"></i><span> Tags</span></a></div><div class="menus_item"><a class="site-page" href="/categories/"><i class="fa-fw fas fa-folder-open"></i><span> Categories</span></a></div><div class="menus_item"><a class="site-page group" href="javascript:void(0);"><i class="fa-fw fas fa-list"></i><span> List</span><i class="fas fa-chevron-down"></i></a><ul class="menus_item_child"><li><a class="site-page child" href="/gallery/"><i class="fa-fw fas fa-images"></i><span> Gallery</span></a></li></ul></div><div class="menus_item"><a class="site-page" href="/link/"><i class="fa-fw fas fa-link"></i><span> Link</span></a></div><div class="menus_item"><a class="site-page" href="/about/"><i class="fa-fw fas fa-heart"></i><span> About</span></a></div></div><div id="toggle-menu"><a class="site-page"><i class="fas fa-bars fa-fw"></i></a></div></div></nav><div id="post-info"><h1 class="post-title">FACETS Testing CNV</h1><div id="post-meta"><div class="meta-firstline"><span class="post-meta-date"><i class="far fa-calendar-alt fa-fw post-meta-icon"></i><span class="post-meta-label">发表于</span><time class="post-meta-date-created" datetime="2020-06-08T16:00:00.000Z" title="发表于 2020-06-09 00:00:00">2020-06-09</time><span class="post-meta-separator">|</span><i class="fas fa-history fa-fw post-meta-icon"></i><span class="post-meta-label">更新于</span><time class="post-meta-date-updated" datetime="2022-09-06T10:11:41.322Z" title="更新于 2022-09-06 18:11:41">2022-09-06</time></span></div><div class="meta-secondline"><span class="post-meta-separator">|</span><span class="post-meta-pv-cv" id="" data-flag-title="FACETS Testing CNV"><i class="far fa-eye fa-fw post-meta-icon"></i><span class="post-meta-label">阅读量:</span><span id="busuanzi_value_page_pv"><i class="fa-solid fa-spinner fa-spin"></i></span></span></div></div></div></header><main class="layout" id="content-inner"><div id="post"><article class="post-content" id="article-container"><h1 id="FACETS"><a href="#FACETS" class="headerlink" title="FACETS"></a>FACETS</h1><p>facets is CNV detection software with using the ascn principle </p>
<h1 id="input"><a href="#input" class="headerlink" title="input"></a>input</h1><p>first Let’s look at the test data given </p>
<p><img src="/img/facets/1591690799039.png" alt="1591690799039"></p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">Chromosome Chromosome of the SNP</span><br><span class="line">Position Position of the SNP</span><br><span class="line">File1R Read depth supporting the REF allele in normal sample</span><br><span class="line">File1A Read depth supporting the ALT allele in normal sample</span><br><span class="line">File2R Read depth supporting the REF allele in tumour sample</span><br><span class="line">File2A Read depth supporting the ALT allele in tumour sample</span><br></pre></td></tr></table></figure>
<p>The software comes with a comparison program, as shown in IGV</p>
<p><img src="/img/facets/1591691071264.png" alt="1591691071264.png"><br>unlike other software ,facets calulates the depth of snp in each bin ,normal CNV software only calculates the depth of each bin then using CBS process.</p>
<h3 id="next-it-Count-whether-the-sum-of-ref-and-alt-of-each-snp-is-heterozygous"><a href="#next-it-Count-whether-the-sum-of-ref-and-alt-of-each-snp-is-heterozygous" class="headerlink" title="next it Count whether the sum of ref and alt of each snp is heterozygous"></a>next it Count whether the sum of ref and alt of each snp is heterozygous</h3><p><img src="/img/facets/1591691653200.png" alt="1591691653200.png"></p>
<h3 id="the-result-is"><a href="#the-result-is" class="headerlink" title="the result is"></a>the result is</h3><p><img src="/img/facets/1591691763518.png" alt="1591691763518.png"></p>
<h2 id="summary"><a href="#summary" class="headerlink" title="summary"></a>summary</h2><p>$1R = REF-N$</p>
<p>$1A = ALT-N$</p>
<p>$2R = REF-T$</p>
<p>$2A = ALT-T$</p>
<h3 id="NEXT"><a href="#NEXT" class="headerlink" title="NEXT"></a>NEXT</h3><p>$1R=countN=N-RD$</p>
<p>$2R=countT=T-RD$</p>
<p>$N-DP=1R+1A$</p>
<p>$T-DP=2R+2A$</p>
<p>$varT=\frac {T-RD} {T-DP}$</p>
<p>$varN=\frac {N-RD} {N-DP}$</p>
<h3 id="as-the-picture-shows"><a href="#as-the-picture-shows" class="headerlink" title="as the picture shows"></a>as the picture shows</h3><p><img src="/img/facets/1591692248504.png" alt="1591692248504.png"></p>
<p>Then the software starts to calculate logR and logOR and gcbias<br>the gcbias plaes look this blog <a target="_blank" rel="noopener" href="http://www.zxzyl.com/archives/988">http://www.zxzyl.com/archives/988</a></p>
<p>the logR is<br>    $ncount &lt;- tapply(rCountN, gcpct, sum)$</p>
<pre><code>$tcount &lt;- tapply(rCountT, gcpct, sum)$

$tsc1=\frac &#123;sum(ncount)&#125; &#123;sum(tcount)&#125;$

$log2(1+rCountT*tscl) - log2(1+rCountN) - gcbias$
</code></pre><p>the logOR is<br>    $1 = varT x countT$</p>
<pre><code>$2 = (1-varT) x countT $

$3 = varN x countN$

$4 = (1-varN) x countN$

than 
## log-odds-ratio (Haldane correction)

$logOR = log(1+0.5)-log(2+0.5)-log(3+0.5)+log(4+0.5) $
## variance of log-odds-ratio (Haldane; Gart &amp; Zweifel Biometrika 1967)

$logOR = (\frac &#123;1&#125; &#123;( [,1]+0.5)&#125; + \frac &#123;1&#125; &#123;( [,2]+0.5)&#125; + \frac &#123;1&#125; &#123;( [,3]+0.5)&#125; + \frac &#123;1&#125; &#123;( [,4]+0.5))&#125;$
</code></pre><h3 id="The-following-steps-are-the-EM-algorithm-here-they-use-Bayesian-in-step-E-to-get-the-posterior-probability"><a href="#The-following-steps-are-the-EM-algorithm-here-they-use-Bayesian-in-step-E-to-get-the-posterior-probability" class="headerlink" title="The following steps are the EM algorithm, here they use Bayesian in step E to get the posterior probability,"></a>The following steps are the EM algorithm, here they use Bayesian in step E to get the posterior probability,</h3><figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br><span class="line">70</span><br><span class="line">71</span><br><span class="line">72</span><br><span class="line">73</span><br></pre></td><td class="code"><pre><span class="line">####LogR mixture model parameter####</span><br><span class="line">    gamma=2</span><br><span class="line">    phi=2*(1-rho)+gamma*rho</span><br><span class="line">    mu=log2(2*(1-rhov)+matrix(rhov,ncol=1)%*%t)-log2(phi)</span><br><span class="line"></span><br><span class="line">    ####LogOR mixture model parameter####</span><br><span class="line">    #allelic ratio</span><br><span class="line">    k=(matrix(rhov,ncol=1)%*%major+1-rhov)/(matrix(rhov,ncol=1)%*%minor+1-rhov)</span><br><span class="line">    logk=log(k)</span><br><span class="line">    logk2=logk^2</span><br><span class="line"></span><br><span class="line">    #posterior probability matrix</span><br><span class="line">    #pmatrix=NULL</span><br><span class="line">    pmatrix=matrix(NA,nrow=nrow(jointseg),ncol=ng)</span><br><span class="line">    loglik=0</span><br><span class="line"></span><br><span class="line">    clust=rep(segclust,nmark)</span><br><span class="line">    segc=sort(unique(segclust[chr&lt;=nX]))</span><br><span class="line">    for(s in segc)&#123;</span><br><span class="line">      idx=which(clust==s)</span><br><span class="line">      x1ij=logR.adj[idx]</span><br><span class="line">      upper=quantile(x1ij,0.95)</span><br><span class="line">      lower=quantile(x1ij,0.05)</span><br><span class="line">      x1ij[x1ij&gt;upper]=NA</span><br><span class="line">      x1ij[x1ij&lt;lower]=NA</span><br><span class="line">      mus=rep(mu[s,],each=length(idx))</span><br><span class="line">      sd=sigma[s]</span><br><span class="line">      if(rhov[s]&lt;0.4)&#123;</span><br><span class="line">        x1ij=rep(cnlr.median.clust[s]-dipLogR,length(idx))</span><br><span class="line">        sd=0.1</span><br><span class="line">        &#125;</span><br><span class="line">      #density for logR.adj (centered logR)</span><br><span class="line">      d1=dnorm(x1ij,mean=mus,sd=sd)</span><br><span class="line">      d1[d1==Inf]=NA</span><br><span class="line"></span><br><span class="line">      #density for logOR, non-central chi-square</span><br><span class="line">      nu=rep(logk2[s,],each=length(idx))</span><br><span class="line">      lambda=nu/rep(logORvar[idx],ng)</span><br><span class="line">      x2ij=logOR2var[idx]</span><br><span class="line">      if(rhov[s]&lt;0.4)&#123;</span><br><span class="line">        x2ij=rep(mafR.clust[s]/logORvar.clust[s],length(idx))</span><br><span class="line">        lambda=nu/logORvar.clust[s]</span><br><span class="line">        &#125;</span><br><span class="line">      #d2=dchisq(x2ij+1,df=1,ncp=lambda)</span><br><span class="line">      d2=dchisq(x2ij,df=1,ncp=lambda)</span><br><span class="line">      d2=1/(abs(x2ij-lambda)+1e-6)</span><br><span class="line">      d2[d2==Inf]=NA</span><br><span class="line"></span><br><span class="line">      #likelihood</span><br><span class="line">      d=d1*d2</span><br><span class="line">      hetsum=d[rep(het[idx]==1,ng)]</span><br><span class="line">      homsum=d1[rep(het[idx]==0,ng)]</span><br><span class="line">      d=sum(hetsum[hetsum&lt;Inf],na.rm=T)+sum(homsum[homsum&lt;Inf],na.rm=T)</span><br><span class="line">      if(!is.na(d)&amp;d&gt;0&amp;d&lt;Inf)&#123;loglik=loglik+log(d)&#125;</span><br><span class="line"></span><br><span class="line">      #heterozygous positions contribute to logR and logOR</span><br><span class="line">      numerator1=matrix(d1*d2,nrow=length(idx),ncol=ng,byrow=F)</span><br><span class="line">      numerator1=sweep(numerator1,MARGIN=2,prior[s,],`*`)</span><br><span class="line"></span><br><span class="line">      #homozygous positions contribute to logR only</span><br><span class="line">      numerator0=matrix(d1,nrow=length(idx),ncol=ng,byrow=F)</span><br><span class="line">      numerator0=sweep(numerator0,MARGIN=2,prior[s,],`*`)</span><br><span class="line"></span><br><span class="line">      numerator=numerator1</span><br><span class="line">      numerator[het[idx]==0,]=numerator0[het[idx]==0,]</span><br><span class="line"></span><br><span class="line">      tmp=apply(numerator,1,function(x)x/(sum(x,na.rm=T)+1e-5))</span><br><span class="line">      #pmatrix=rbind(pmatrix,t(tmp))</span><br><span class="line">      pmatrix[idx,]=t(tmp)</span><br><span class="line"></span><br><span class="line">      #update prior</span><br><span class="line">      prior[s,]=apply(t(tmp),2,function(x)mean(x,na.rm=T))</span><br><span class="line">    &#125;</span><br></pre></td></tr></table></figure>
<p><img src="/img/facets/1591695237945.png" alt="1591695237945.png"></p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br><span class="line">70</span><br><span class="line">71</span><br><span class="line">72</span><br><span class="line">73</span><br><span class="line">74</span><br><span class="line">75</span><br><span class="line">76</span><br><span class="line">77</span><br><span class="line">78</span><br><span class="line">79</span><br><span class="line">80</span><br><span class="line">81</span><br><span class="line">82</span><br><span class="line">83</span><br><span class="line">84</span><br><span class="line">85</span><br><span class="line">86</span><br><span class="line">87</span><br><span class="line">88</span><br><span class="line">89</span><br><span class="line">90</span><br><span class="line">91</span><br><span class="line">92</span><br><span class="line">93</span><br><span class="line">94</span><br><span class="line">95</span><br></pre></td><td class="code"><pre><span class="line">###  the M step is </span><br><span class="line">#get CF per segments, pick mode close to 1 (favor high purity low cn solution)</span><br><span class="line">    rhom=gammam=matrix(NA,nrow=nclust,ncol=ng)</span><br><span class="line">    geno=matrix(0,nrow=nclust,ncol=ng)</span><br><span class="line">    which.geno=posterior=rep(NA,nclust)</span><br><span class="line">    for(i in segc)&#123;</span><br><span class="line"></span><br><span class="line">      idx=which(clust==i)</span><br><span class="line">      idxhet=which(clust==i&amp;het==1)</span><br><span class="line">      sump=apply(pmatrix[idx,,drop=F],2,function(x)sum(x,na.rm=T))</span><br><span class="line"></span><br><span class="line">      #if probability is too small (highly uncertain), use lsd estimates for stability </span><br><span class="line">      if(all(is.na(prior[i,])))&#123;</span><br><span class="line">        prior[i,]=prior.old[i,]</span><br><span class="line">        &#125;else&#123;</span><br><span class="line">        if(sum(prior[i,],na.rm=T)==0)prior[i,]=prior.old[i,]</span><br><span class="line">        &#125;</span><br><span class="line"></span><br><span class="line">      if(max(prior[i,],na.rm=T)&gt;0.05)&#123;   </span><br><span class="line"></span><br><span class="line">      ##calculate rho for the most likely genotype(s) for segment i</span><br><span class="line">      ##if there more more than one likely candidates save two and pick one with higher CF</span><br><span class="line">      #top2=sort(prior[i,],decreasing=T)[1:2]</span><br><span class="line">      #if(top2[2]&gt;0.05&amp;abs(diff(top2))&lt;0.0001)&#123;</span><br><span class="line">      #    sump[prior[i,]&lt;quantile(prior[i,],(ng-2)/ng)]=NA</span><br><span class="line">      # &#125;else&#123;</span><br><span class="line">      #    sump[prior[i,]&lt;max(prior[i,])]=NA</span><br><span class="line">      # &#125;</span><br><span class="line"></span><br><span class="line">      sump[prior[i,]&lt;max(prior[i,])]=NA</span><br><span class="line"></span><br><span class="line">        ##update k</span><br><span class="line">        tmphet=pmatrix[idxhet,,drop=F]</span><br><span class="line">        v1=as.vector((logOR[idxhet]^2-logORvar[idxhet])/logORvar[idxhet])</span><br><span class="line">        v2=as.vector(1/logORvar[idxhet])</span><br><span class="line">        sumdphet=apply(sweep(tmphet,MARGIN=1, v1, `*`), 2,function(x)sum(x,na.rm=T))</span><br><span class="line">        sumphet=apply(sweep(tmphet,MARGIN=1,v2,`*`), 2,function(x)sum(x,na.rm=T))</span><br><span class="line">        sumphet[is.na(sump)]=NA</span><br><span class="line"></span><br><span class="line">        #CF from logOR    </span><br><span class="line">        logk2hat=pmax(0,sumdphet/sumphet) #can be negative when k=1 logk=0 set to 0</span><br><span class="line">        khat=exp(sqrt(logk2hat))</span><br><span class="line">        a=(1-khat)/(khat*(minor-1)-(major-1))</span><br><span class="line">        a[abs(a)==Inf]=NA</span><br><span class="line">        a[a&lt;=0]=NA</span><br><span class="line">        a[a&gt;1]=1</span><br><span class="line">        if(all(nhet[segclust==i]&lt;min.nhet))a=rep(NA,ng)</span><br><span class="line"></span><br><span class="line">        #CF from logR</span><br><span class="line">        tmp=pmatrix[idx,,drop=F]</span><br><span class="line">        v=as.vector(logR.adj[idx])</span><br><span class="line">        sumdp=apply(sweep(tmp,MARGIN=1,v,`*`),2,function(x)sum(x,na.rm=T))   </span><br><span class="line">        mu.hat=sumdp/sump #mu.hat</span><br><span class="line">        aa=2*(2^mu.hat-1)/(t-2)</span><br><span class="line">        aa[abs(aa)==Inf]=NA</span><br><span class="line">        aa[aa&lt;=0]=NA</span><br><span class="line">        aa[aa&gt;1]=1</span><br><span class="line"></span><br><span class="line">        aaa=pmax(a,aa,na.rm=T)</span><br><span class="line">        #degenerate cases</span><br><span class="line">        #homozygous deletion (0) and balanced gain (AABB, AAABBB), maf=0.5, purity information comes from logr only</span><br><span class="line">        aaa[c(1,8,13)]=aa[c(1,8,13)]  </span><br><span class="line"></span><br><span class="line">        #set upper bound at sample rho</span><br><span class="line">        aaa=pmin(aaa,rho)</span><br><span class="line"></span><br><span class="line">        #uniparental disomy (AA) CF information comes from logOR only.</span><br><span class="line"></span><br><span class="line">        ##if there are two likely genotype, choose one with higher purity (e.g.,AAB 80% or AAAB 50%)</span><br><span class="line">        ##if the higher CF exceeds sample purity, then the lower CF is the right one</span><br><span class="line">        #if(all(is.na(aaa)))&#123;which.geno[i]=which.max(prior[i,])&#125;else&#123;</span><br><span class="line">        #  which.geno[i]=ifelse(max(aaa,na.rm=T)&lt;rho,which.max(aaa),which.min(aaa))</span><br><span class="line">        #&#125;</span><br><span class="line"></span><br><span class="line">        which.geno[i]=which.max(prior[i,])</span><br><span class="line"></span><br><span class="line">        postprob=pmatrix[idx,which.geno[i]]</span><br><span class="line">        posterior[i]=mean(postprob[postprob&gt;0],na.rm=T)</span><br><span class="line"></span><br><span class="line">        #update sigma</span><br><span class="line">        y=as.vector(logR.adj[idx])*pmatrix[idx,,drop=F]</span><br><span class="line">        r=y-mu[i,]*pmatrix[idx,,drop=F]</span><br><span class="line">        ss=sqrt(sum(r[,which.geno[i]]^2,na.rm=T)/sum(pmatrix[idx,which.geno[i]])) </span><br><span class="line">        sigma[i]=ifelse(is.na(ss),0.5,ss)</span><br><span class="line"></span><br><span class="line">        aaa[setdiff(1:ng,which.geno[i])]=NA  </span><br><span class="line"></span><br><span class="line">        #het dip (AB) seg has no information, set CF at a high value less than 1</span><br><span class="line">        #if(any(which(!is.na(sump))==4))&#123;aaa[4]=0.9&#125;</span><br><span class="line"></span><br><span class="line">        rhom[i,]=aaa </span><br><span class="line"></span><br><span class="line">      &#125; #max prior</span><br><span class="line"></span><br><span class="line">    &#125;</span><br></pre></td></tr></table></figure>
<p><img src="/img/facets/1591695253756.png" alt="1591695253756.png"></p>
<h3 id="Loop-until-it-converges"><a href="#Loop-until-it-converges" class="headerlink" title="Loop until it converges"></a>Loop until it converges</h3><h2 id="Or-simpler-also-use-the-CBS-algorithm-instead-it"><a href="#Or-simpler-also-use-the-CBS-algorithm-instead-it" class="headerlink" title="Or simpler, also use the CBS algorithm instead it"></a>Or simpler, also use the CBS algorithm instead it</h2><h1 id="picard方式计算CNV"><a href="#picard方式计算CNV" class="headerlink" title="picard方式计算CNV"></a>picard方式计算CNV</h1><p>探针长度</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">chrom   start   end     length  </span><br><span class="line">chr1    12080   12251   172 </span><br><span class="line">chr1    12595   12802   208</span><br></pre></td></tr></table></figure>
<p>normalized_coverage : for each target interval, the read depth (unique read starts) that correspond to a particular target interval is divided by the average number of read starts in all of the target intervals.</p>
<h4 id="normal"><a href="#normal" class="headerlink" title="normal"></a>normal</h4><figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">chrom   start   end     length   normalized_coverage    gene</span><br><span class="line">chr1    12080   12251   172        0                    DDX11L1</span><br><span class="line">chr1    12595   12802   208        0.002957             DDX11L1</span><br></pre></td></tr></table></figure>
<h4 id="tumor"><a href="#tumor" class="headerlink" title="tumor"></a>tumor</h4><figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">chrom   start   end     length   normalized_coverage   gene</span><br><span class="line">chr1    12080   12251   172        0                   DDX11L1</span><br><span class="line">chr1    12595   12802   208        0.011583            DDX11L1</span><br></pre></td></tr></table></figure>
<h3 id="算法"><a href="#算法" class="headerlink" title="算法"></a>算法</h3><p>例如gene  EGFR</p>
<p>片段长度</p>
<p>$length = L1 L2  L3  L4  L5$</p>
<p>归一化覆盖:</p>
<p>$Tn = Tn<em>{1}\ Tn</em>{2}\ Tn<em>{3}\ Tn</em>{4}\ Tn_{5}$</p>
<p>$Nn = Nn<em>{1}\ Nn</em>{2}\ Nn<em>{3}\ Nn</em>{4}\ Nn_{5}$</p>
<p>计算 :<br>$\frac {(L1(\frac {Tn<em>{1}} {Nn</em>{1}})+L2(\frac {Tn<em>{2}} {Nn</em>{2}})+·····)} {L1+···L5} =copy\ number\ ratio$</p>
</article><div class="post-copyright"><div class="post-copyright__author"><span class="post-copyright-meta">文章作者: </span><span class="post-copyright-info"><a href="https://zhaohongqiangsoliva.github.io">Solivehong</a></span></div><div class="post-copyright__type"><span class="post-copyright-meta">文章链接: </span><span class="post-copyright-info"><a href="https://zhaohongqiangsoliva.github.io/2020/06/09/2020-06-09-CNV-Facets%E8%BD%AF%E4%BB%B6%E6%A3%80%E6%B5%8Bcnv/">https://zhaohongqiangsoliva.github.io/2020/06/09/2020-06-09-CNV-Facets%E8%BD%AF%E4%BB%B6%E6%A3%80%E6%B5%8Bcnv/</a></span></div><div class="post-copyright__notice"><span class="post-copyright-meta">版权声明: </span><span class="post-copyright-info">本博客所有文章除特别声明外，均采用 <a href="https://creativecommons.org/licenses/by-nc-sa/4.0/" target="_blank">CC BY-NC-SA 4.0</a> 许可协议。转载请注明来自 <a href="https://zhaohongqiangsoliva.github.io" target="_blank">Solivehong Blog</a>！</span></div></div><div class="tag_share"><div class="post-meta__tag-list"><a class="post-meta__tags" href="/tags/%E7%94%9F%E4%BF%A1/">生信</a></div><div class="post_share"><div class="social-share" data-image="/img/dna.png" data-sites="facebook,twitter,wechat,weibo,qq"></div><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/butterfly-extsrc/sharejs/dist/css/share.min.css" media="print" onload="this.media='all'"><script src="https://cdn.jsdelivr.net/npm/butterfly-extsrc/sharejs/dist/js/social-share.min.js" defer></script></div></div><nav class="pagination-post" id="pagination"><div class="prev-post pull-left"><a href="/2020/07/13/2020-07-13-%E7%94%9F%E4%BF%A1%E5%9F%B9%E8%AE%AD%E6%89%8B%E5%86%8C/"><img class="prev-cover" src="/" onerror="onerror=null;src='/img/404.jpg'" alt="cover of previous post"><div class="pagination-info"><div class="label">上一篇</div><div class="prev_info">生信培训手册</div></div></a></div><div class="next-post pull-right"><a href="/2020/06/06/2020-06-06-T%E6%A3%80%E9%AA%8C-%E8%AE%A1%E7%AE%97TMB%E4%B8%A4%E7%A7%8D%E6%96%B9%E5%BC%8F%E5%B7%AE%E5%BC%82/"><img class="next-cover" src="/" onerror="onerror=null;src='/img/404.jpg'" alt="cover of next post"><div class="pagination-info"><div class="label">下一篇</div><div class="next_info">T检验-计算TMB两种方式差异</div></div></a></div></nav><div class="relatedPosts"><div class="headline"><i class="fas fa-thumbs-up fa-fw"></i><span>相关推荐</span></div><div class="relatedPosts-list"><div><a href="/2020/06/06/2020-06-06-T%E6%A3%80%E9%AA%8C-%E8%AE%A1%E7%AE%97TMB%E4%B8%A4%E7%A7%8D%E6%96%B9%E5%BC%8F%E5%B7%AE%E5%BC%82/" title="T检验-计算TMB两种方式差异"><img class="cover" src="/" alt="cover"><div class="content is-center"><div class="date"><i class="far fa-calendar-alt fa-fw"></i> 2020-06-06</div><div class="title">T检验-计算TMB两种方式差异</div></div></a></div><div><a href="/2020/06/06/2020-06-06-T%E6%A3%80%E9%AA%8C%E3%80%81F%E6%A3%80%E9%AA%8C%E3%80%81%E5%8D%A1%E6%96%B9%E6%A3%80%E9%AA%8C%E8%AF%A6%E7%BB%86%E5%88%86%E6%9E%90%E5%8F%8A%E5%BA%94%E7%94%A8%E5%9C%BA%E6%99%AF%E6%80%BB%E7%BB%93/" title="T检验、F检验、卡方检验详细分析及应用场景总结"><img class="cover" src="/" alt="cover"><div class="content is-center"><div class="date"><i class="far fa-calendar-alt fa-fw"></i> 2020-06-06</div><div class="title">T检验、F检验、卡方检验详细分析及应用场景总结</div></div></a></div><div><a href="/2020/06/06/2020-06-06-%E9%A9%B1%E5%8A%A8%E5%9F%BA%E5%9B%A0%E6%A3%80%E6%B5%8B/" title="驱动基因"><img class="cover" src="/" alt="cover"><div class="content is-center"><div class="date"><i class="far fa-calendar-alt fa-fw"></i> 2020-06-06</div><div class="title">驱动基因</div></div></a></div><div><a href="/2020/07/13/2020-07-13-%E7%94%9F%E4%BF%A1%E5%9F%B9%E8%AE%AD%E6%89%8B%E5%86%8C/" title="生信培训手册"><img class="cover" src="/" alt="cover"><div class="content is-center"><div class="date"><i class="far fa-calendar-alt fa-fw"></i> 2020-07-13</div><div class="title">生信培训手册</div></div></a></div><div><a href="/2020/07/14/2020-08-11-%E4%BD%93%E7%BB%86%E8%83%9Ecalling%20%E7%AE%97%E6%B3%95%E7%BB%BC%E8%BF%B0/" title="体细胞calling 算法综述"><img class="cover" src="/" alt="cover"><div class="content is-center"><div class="date"><i class="far fa-calendar-alt fa-fw"></i> 2020-07-14</div><div class="title">体细胞calling 算法综述</div></div></a></div></div></div></div><div class="aside-content" id="aside-content"><div class="card-widget card-info"><div class="is-center"><div class="avatar-img"><img src="/img/dna.png" onerror="this.onerror=null;this.src='/img/friend_404.gif'" alt="avatar"/></div><div class="author-info__name">Solivehong</div><div class="author-info__description"></div></div><div class="card-info-data site-data is-center"><a href="/archives/"><div class="headline">文章</div><div class="length-num">19</div></a><a href="/tags/"><div class="headline">标签</div><div class="length-num">4</div></a><a href="/categories/"><div class="headline">分类</div><div class="length-num">0</div></a></div><a id="card-info-btn" target="_blank" rel="noopener" href="https://github.com/xxxxxx"><i class="fab fa-github"></i><span>Follow Me</span></a><div class="card-info-social-icons is-center"><a class="social-icon" href="https://github.com/zhaohongqiangsoliva" target="_blank" title="Github"><i class="fab fa-github"></i></a><a class="social-icon" href="mailto:zhaohongqiangsoliva@gmail.com" target="_blank" title="Email"><i class="fas fa-envelope"></i></a></div></div><div class="card-widget card-announcement"><div class="item-headline"><i class="fas fa-bullhorn fa-shake"></i><span>公告</span></div><div class="announcement_content">This is my Blog</div></div><div class="sticky_layout"><div class="card-widget" id="card-toc"><div class="item-headline"><i class="fas fa-stream"></i><span>目录</span><span class="toc-percentage"></span></div><div class="toc-content"><ol class="toc"><li class="toc-item toc-level-1"><a class="toc-link" href="#FACETS"><span class="toc-number">1.</span> <span class="toc-text">FACETS</span></a></li><li class="toc-item toc-level-1"><a class="toc-link" href="#input"><span class="toc-number">2.</span> <span class="toc-text">input</span></a><ol class="toc-child"><li class="toc-item toc-level-3"><a class="toc-link" href="#next-it-Count-whether-the-sum-of-ref-and-alt-of-each-snp-is-heterozygous"><span class="toc-number">2.0.1.</span> <span class="toc-text">next it Count whether the sum of ref and alt of each snp is heterozygous</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#the-result-is"><span class="toc-number">2.0.2.</span> <span class="toc-text">the result is</span></a></li></ol></li><li class="toc-item toc-level-2"><a class="toc-link" href="#summary"><span class="toc-number">2.1.</span> <span class="toc-text">summary</span></a><ol class="toc-child"><li class="toc-item toc-level-3"><a class="toc-link" href="#NEXT"><span class="toc-number">2.1.1.</span> <span class="toc-text">NEXT</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#as-the-picture-shows"><span class="toc-number">2.1.2.</span> <span class="toc-text">as the picture shows</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#The-following-steps-are-the-EM-algorithm-here-they-use-Bayesian-in-step-E-to-get-the-posterior-probability"><span class="toc-number">2.1.3.</span> <span class="toc-text">The following steps are the EM algorithm, here they use Bayesian in step E to get the posterior probability,</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#Loop-until-it-converges"><span class="toc-number">2.1.4.</span> <span class="toc-text">Loop until it converges</span></a></li></ol></li><li class="toc-item toc-level-2"><a class="toc-link" href="#Or-simpler-also-use-the-CBS-algorithm-instead-it"><span class="toc-number">2.2.</span> <span class="toc-text">Or simpler, also use the CBS algorithm instead it</span></a></li></ol></li><li class="toc-item toc-level-1"><a class="toc-link" href="#picard%E6%96%B9%E5%BC%8F%E8%AE%A1%E7%AE%97CNV"><span class="toc-number">3.</span> <span class="toc-text">picard方式计算CNV</span></a><ol class="toc-child"><li class="toc-item toc-level-4"><a class="toc-link" href="#normal"><span class="toc-number">3.0.0.1.</span> <span class="toc-text">normal</span></a></li><li class="toc-item toc-level-4"><a class="toc-link" href="#tumor"><span class="toc-number">3.0.0.2.</span> <span class="toc-text">tumor</span></a></li></ol></li><li class="toc-item toc-level-3"><a class="toc-link" href="#%E7%AE%97%E6%B3%95"><span class="toc-number">3.0.1.</span> <span class="toc-text">算法</span></a></li></ol></li></ol></li></ol></div></div><div class="card-widget card-recent-post"><div class="item-headline"><i class="fas fa-history"></i><span>最新文章</span></div><div class="aside-list"><div class="aside-list-item"><a class="thumbnail" href="/2022/10/27/2022-10-27-linuxTools%E6%89%8B%E5%86%8C%E5%A4%A7%E5%85%A8/" title="无题"><img src="https://unpkg.com/justlovesmile-img/cover4.jpg" onerror="this.onerror=null;this.src='/img/404.jpg'" alt="无题"/></a><div class="content"><a class="title" href="/2022/10/27/2022-10-27-linuxTools%E6%89%8B%E5%86%8C%E5%A4%A7%E5%85%A8/" title="无题">无题</a><time datetime="2022-10-27T13:13:28.638Z" title="发表于 2022-10-27 21:13:28">2022-10-27</time></div></div><div class="aside-list-item"><a class="thumbnail" href="/2022/10/27/2022-10-27-fzf%E7%9A%84%E4%BD%BF%E7%94%A8/" title="无题"><img src="https://unpkg.com/justlovesmile-img/cover6.JPG" onerror="this.onerror=null;this.src='/img/404.jpg'" alt="无题"/></a><div class="content"><a class="title" href="/2022/10/27/2022-10-27-fzf%E7%9A%84%E4%BD%BF%E7%94%A8/" title="无题">无题</a><time datetime="2022-10-27T12:40:30.381Z" title="发表于 2022-10-27 20:40:30">2022-10-27</time></div></div><div class="aside-list-item"><div class="content"><a class="title" href="/2022/09/04/2022-09-04-Mixed_linear_model/" title="无题">无题</a><time datetime="2022-09-04T08:58:03.456Z" title="发表于 2022-09-04 16:58:03">2022-09-04</time></div></div><div class="aside-list-item"><div class="content"><a class="title" href="/2022/08/27/2022-08-26-oh-my-bash/" title="无题">无题</a><time datetime="2022-08-27T03:23:41.553Z" title="发表于 2022-08-27 11:23:41">2022-08-27</time></div></div><div class="aside-list-item"><div class="content"><a class="title" href="/2022/07/09/2021-10-02-%E5%BC%BA%E5%8C%96%E5%AD%A6%E4%B9%A0-01-MDPs/" title="无题">无题</a><time datetime="2022-07-09T03:00:19.114Z" title="发表于 2022-07-09 11:00:19">2022-07-09</time></div></div></div></div></div></div></main><footer id="footer"><div id="footer-wrap"><div class="copyright">&copy;2020 - 2022 By Solivehong</div><div class="framework-info"><span>框架 </span><a target="_blank" rel="noopener" href="https://hexo.io">Hexo</a><span class="footer-separator">|</span><span>主题 </span><a target="_blank" rel="noopener" href="https://github.com/jerryc127/hexo-theme-butterfly">Butterfly</a></div></div></footer></div><div id="rightside"><div id="rightside-config-hide"><button id="readmode" type="button" title="阅读模式"><i class="fas fa-book-open"></i></button><button id="darkmode" type="button" title="浅色和深色模式转换"><i class="fas fa-adjust"></i></button><button id="hide-aside-btn" type="button" title="单栏和双栏切换"><i class="fas fa-arrows-alt-h"></i></button></div><div id="rightside-config-show"><button id="rightside_config" type="button" title="设置"><i class="fas fa-cog fa-spin"></i></button><button class="close" id="mobile-toc-button" type="button" title="目录"><i class="fas fa-list-ul"></i></button><button id="go-up" type="button" title="回到顶部"><i class="fas fa-arrow-up"></i></button></div></div><div id="local-search"><div class="search-dialog"><nav class="search-nav"><span class="search-dialog-title">搜索</span><span id="loading-status"></span><button class="search-close-button"><i class="fas fa-times"></i></button></nav><div class="is-center" id="loading-database"><i class="fas fa-spinner fa-pulse"></i><span>  数据库加载中</span></div><div class="search-wrap"><div id="local-search-input"><div class="local-search-box"><input class="local-search-box--input" placeholder="搜索文章" type="text"/></div></div><hr/><div id="local-search-results"></div></div></div><div id="search-mask"></div></div><div><script src="/js/utils.js"></script><script src="/js/main.js"></script><script src="https://cdn.jsdelivr.net/npm/@fancyapps/ui/dist/fancybox.umd.min.js"></script><script src="/js/search/local-search.js"></script><div class="js-pjax"><script>if (!window.MathJax) {
  window.MathJax = {
    tex: {
      inlineMath: [ ['$','$'], ["\\(","\\)"]],
      tags: 'ams'
    },
    chtml: {
      scale: 1.2
    },
    options: {
      renderActions: {
        findScript: [10, doc => {
          for (const node of document.querySelectorAll('script[type^="math/tex"]')) {
            const display = !!node.type.match(/; *mode=display/)
            const math = new doc.options.MathItem(node.textContent, doc.inputJax[0], display)
            const text = document.createTextNode('')
            node.parentNode.replaceChild(text, node)
            math.start = {node: text, delim: '', n: 0}
            math.end = {node: text, delim: '', n: 0}
            doc.math.push(math)
          }
        }, ''],
        insertScript: [200, () => {
          document.querySelectorAll('mjx-container:not\([display]\)').forEach(node => {
            const target = node.parentNode
            if (target.nodeName.toLowerCase() === 'li') {
              target.parentNode.classList.add('has-jax')
            } else {
              target.classList.add('has-jax')
            }
          });
        }, '', false]
      }
    }
  }
  
  const script = document.createElement('script')
  script.src = 'https://cdn.jsdelivr.net/npm/mathjax/es5/tex-mml-chtml.min.js'
  script.id = 'MathJax-script'
  script.async = true
  document.head.appendChild(script)
} else {
  MathJax.startup.document.state(0)
  MathJax.texReset()
  MathJax.typeset()
}</script></div><script async data-pjax src="//busuanzi.ibruce.info/busuanzi/2.3/busuanzi.pure.mini.js"></script></div></body></html>