<!DOCTYPE html><html lang="zh-CN" data-theme="light"><head><meta charset="UTF-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no"><title>体细胞calling 算法综述 | Solivehong Blog</title><meta name="keywords" content="生信"><meta name="author" content="Solivehong"><meta name="copyright" content="Solivehong"><meta name="format-detection" content="telephone=no"><meta name="theme-color" content="ffffff"><meta name="description" content="体细胞calling 算法综述 somatic single nucleotide variant calling algorithmspaper基本信息paper name：A review of somatic single nucleotide variant calling algorithms for next-generation sequencing datadoi :10.1016">
<meta property="og:type" content="article">
<meta property="og:title" content="体细胞calling 算法综述">
<meta property="og:url" content="https://zhaohongqiangsoliva.github.io/2020/07/14/2020-08-11-%E4%BD%93%E7%BB%86%E8%83%9Ecalling%20%E7%AE%97%E6%B3%95%E7%BB%BC%E8%BF%B0/index.html">
<meta property="og:site_name" content="Solivehong Blog">
<meta property="og:description" content="体细胞calling 算法综述 somatic single nucleotide variant calling algorithmspaper基本信息paper name：A review of somatic single nucleotide variant calling algorithms for next-generation sequencing datadoi :10.1016">
<meta property="og:locale" content="zh_CN">
<meta property="og:image" content="https://zhaohongqiangsoliva.github.io/img/dna.png">
<meta property="article:published_time" content="2020-07-13T16:00:00.000Z">
<meta property="article:modified_time" content="2020-08-11T22:01:50.989Z">
<meta property="article:author" content="Solivehong">
<meta property="article:tag" content="生信">
<meta name="twitter:card" content="summary">
<meta name="twitter:image" content="https://zhaohongqiangsoliva.github.io/img/dna.png"><link rel="shortcut icon" href="/img/favicon.png"><link rel="canonical" href="https://zhaohongqiangsoliva.github.io/2020/07/14/2020-08-11-%E4%BD%93%E7%BB%86%E8%83%9Ecalling%20%E7%AE%97%E6%B3%95%E7%BB%BC%E8%BF%B0/"><link rel="preconnect" href="//cdn.jsdelivr.net"/><link rel="preconnect" href="//busuanzi.ibruce.info"/><link rel="stylesheet" href="/css/index.css"><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/@fortawesome/fontawesome-free/css/all.min.css" media="print" onload="this.media='all'"><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/@fancyapps/ui/dist/fancybox.min.css" media="print" onload="this.media='all'"><script>const GLOBAL_CONFIG = { 
  root: '/',
  algolia: undefined,
  localSearch: {"path":"/search.xml","preload":true,"languages":{"hits_empty":"找不到您查询的内容：${query}"}},
  translate: undefined,
  noticeOutdate: undefined,
  highlight: {"plugin":"highlighjs","highlightCopy":true,"highlightLang":true,"highlightHeightLimit":false},
  copy: {
    success: '复制成功',
    error: '复制错误',
    noSupport: '浏览器不支持'
  },
  relativeDate: {
    homepage: false,
    post: false
  },
  runtime: '',
  date_suffix: {
    just: '刚刚',
    min: '分钟前',
    hour: '小时前',
    day: '天前',
    month: '个月前'
  },
  copyright: undefined,
  lightbox: 'fancybox',
  Snackbar: undefined,
  source: {
    justifiedGallery: {
      js: 'https://cdn.jsdelivr.net/npm/flickr-justified-gallery/dist/fjGallery.min.js',
      css: 'https://cdn.jsdelivr.net/npm/flickr-justified-gallery/dist/fjGallery.min.css'
    }
  },
  isPhotoFigcaption: false,
  islazyload: false,
  isAnchor: false
}</script><script id="config-diff">var GLOBAL_CONFIG_SITE = {
  title: '体细胞calling 算法综述',
  isPost: true,
  isHome: false,
  isHighlightShrink: false,
  isToc: true,
  postUpdate: '2020-08-12 06:01:50'
}</script><noscript><style type="text/css">
  #nav {
    opacity: 1
  }
  .justified-gallery img {
    opacity: 1
  }

  #recent-posts time,
  #post-meta time {
    display: inline !important
  }
</style></noscript><script>(win=>{
    win.saveToLocal = {
      set: function setWithExpiry(key, value, ttl) {
        if (ttl === 0) return
        const now = new Date()
        const expiryDay = ttl * 86400000
        const item = {
          value: value,
          expiry: now.getTime() + expiryDay,
        }
        localStorage.setItem(key, JSON.stringify(item))
      },

      get: function getWithExpiry(key) {
        const itemStr = localStorage.getItem(key)

        if (!itemStr) {
          return undefined
        }
        const item = JSON.parse(itemStr)
        const now = new Date()

        if (now.getTime() > item.expiry) {
          localStorage.removeItem(key)
          return undefined
        }
        return item.value
      }
    }
  
    win.getScript = url => new Promise((resolve, reject) => {
      const script = document.createElement('script')
      script.src = url
      script.async = true
      script.onerror = reject
      script.onload = script.onreadystatechange = function() {
        const loadState = this.readyState
        if (loadState && loadState !== 'loaded' && loadState !== 'complete') return
        script.onload = script.onreadystatechange = null
        resolve()
      }
      document.head.appendChild(script)
    })
  
      win.activateDarkMode = function () {
        document.documentElement.setAttribute('data-theme', 'dark')
        if (document.querySelector('meta[name="theme-color"]') !== null) {
          document.querySelector('meta[name="theme-color"]').setAttribute('content', '#0d0d0d')
        }
      }
      win.activateLightMode = function () {
        document.documentElement.setAttribute('data-theme', 'light')
        if (document.querySelector('meta[name="theme-color"]') !== null) {
          document.querySelector('meta[name="theme-color"]').setAttribute('content', 'ffffff')
        }
      }
      const t = saveToLocal.get('theme')
    
          if (t === 'dark') activateDarkMode()
          else if (t === 'light') activateLightMode()
        
      const asideStatus = saveToLocal.get('aside-status')
      if (asideStatus !== undefined) {
        if (asideStatus === 'hide') {
          document.documentElement.classList.add('hide-aside')
        } else {
          document.documentElement.classList.remove('hide-aside')
        }
      }
    
    const detectApple = () => {
      if(/iPad|iPhone|iPod|Macintosh/.test(navigator.userAgent)){
        document.documentElement.classList.add('apple')
      }
    }
    detectApple()
    })(window)</script><meta name="generator" content="Hexo 6.2.0"></head><body><div id="web_bg"></div><div id="sidebar"><div id="menu-mask"></div><div id="sidebar-menus"><div class="avatar-img is-center"><img src="/img/dna.png" onerror="onerror=null;src='/img/friend_404.gif'" alt="avatar"/></div><div class="sidebar-site-data site-data is-center"><a href="/archives/"><div class="headline">文章</div><div class="length-num">19</div></a><a href="/tags/"><div class="headline">标签</div><div class="length-num">6</div></a><a href="/categories/"><div class="headline">分类</div><div class="length-num">0</div></a></div><hr/><div class="menus_items"><div class="menus_item"><a class="site-page" href="/"><i class="fa-fw fas fa-home"></i><span> Home</span></a></div><div class="menus_item"><a class="site-page" href="/archives/"><i class="fa-fw fas fa-archive"></i><span> Archives</span></a></div><div class="menus_item"><a class="site-page" href="/tags/"><i class="fa-fw fas fa-tags"></i><span> Tags</span></a></div><div class="menus_item"><a class="site-page" href="/categories/"><i class="fa-fw fas fa-folder-open"></i><span> Categories</span></a></div><div class="menus_item"><a class="site-page group" href="javascript:void(0);"><i class="fa-fw fas fa-list"></i><span> List</span><i class="fas fa-chevron-down"></i></a><ul class="menus_item_child"><li><a class="site-page child" href="/gallery/"><i class="fa-fw fas fa-images"></i><span> Gallery</span></a></li></ul></div><div class="menus_item"><a class="site-page" href="/link/"><i class="fa-fw fas fa-link"></i><span> Link</span></a></div><div class="menus_item"><a class="site-page" href="/about/"><i class="fa-fw fas fa-heart"></i><span> About</span></a></div></div></div></div><div class="post" id="body-wrap"><header class="post-bg" id="page-header"><nav id="nav"><span id="blog_name"><a id="site-name" href="/">Solivehong Blog</a></span><div id="menus"><div id="search-button"><a class="site-page social-icon search"><i class="fas fa-search fa-fw"></i><span> 搜索</span></a></div><div class="menus_items"><div class="menus_item"><a class="site-page" href="/"><i class="fa-fw fas fa-home"></i><span> Home</span></a></div><div class="menus_item"><a class="site-page" href="/archives/"><i class="fa-fw fas fa-archive"></i><span> Archives</span></a></div><div class="menus_item"><a class="site-page" href="/tags/"><i class="fa-fw fas fa-tags"></i><span> Tags</span></a></div><div class="menus_item"><a class="site-page" href="/categories/"><i class="fa-fw fas fa-folder-open"></i><span> Categories</span></a></div><div class="menus_item"><a class="site-page group" href="javascript:void(0);"><i class="fa-fw fas fa-list"></i><span> List</span><i class="fas fa-chevron-down"></i></a><ul class="menus_item_child"><li><a class="site-page child" href="/gallery/"><i class="fa-fw fas fa-images"></i><span> Gallery</span></a></li></ul></div><div class="menus_item"><a class="site-page" href="/link/"><i class="fa-fw fas fa-link"></i><span> Link</span></a></div><div class="menus_item"><a class="site-page" href="/about/"><i class="fa-fw fas fa-heart"></i><span> About</span></a></div></div><div id="toggle-menu"><a class="site-page"><i class="fas fa-bars fa-fw"></i></a></div></div></nav><div id="post-info"><h1 class="post-title">体细胞calling 算法综述</h1><div id="post-meta"><div class="meta-firstline"><span class="post-meta-date"><i class="far fa-calendar-alt fa-fw post-meta-icon"></i><span class="post-meta-label">发表于</span><time class="post-meta-date-created" datetime="2020-07-13T16:00:00.000Z" title="发表于 2020-07-14 00:00:00">2020-07-14</time><span class="post-meta-separator">|</span><i class="fas fa-history fa-fw post-meta-icon"></i><span class="post-meta-label">更新于</span><time class="post-meta-date-updated" datetime="2020-08-11T22:01:50.989Z" title="更新于 2020-08-12 06:01:50">2020-08-12</time></span></div><div class="meta-secondline"><span class="post-meta-separator">|</span><span class="post-meta-pv-cv" id="" data-flag-title="体细胞calling 算法综述"><i class="far fa-eye fa-fw post-meta-icon"></i><span class="post-meta-label">阅读量:</span><span id="busuanzi_value_page_pv"><i class="fa-solid fa-spinner fa-spin"></i></span></span></div></div></div></header><main class="layout" id="content-inner"><div id="post"><article class="post-content" id="article-container"><h1 id="体细胞calling-算法综述-somatic-single-nucleotide-variant-calling-algorithms"><a href="#体细胞calling-算法综述-somatic-single-nucleotide-variant-calling-algorithms" class="headerlink" title="体细胞calling 算法综述 somatic single nucleotide variant calling algorithms"></a>体细胞calling 算法综述 somatic single nucleotide variant calling algorithms</h1><h2 id="paper基本信息"><a href="#paper基本信息" class="headerlink" title="paper基本信息"></a>paper基本信息</h2><p>paper name：<br>A review of somatic single nucleotide variant calling algorithms for next-generation sequencing data<br>doi :10.1016/j.csbj.2018.01.003</p>
<h2 id="总结"><a href="#总结" class="headerlink" title="总结"></a>总结</h2><p>call 变异的主要步骤为 mapping -&gt; call 变异<br>mapping 的DNA工具主要为BWA RNA 工具为star<br>本文着重介绍了目前市面上的call变异的工具的原理<br>算法可以归纳为四类: 匹配肿瘤正常变异、单样本变异、基于 umi 的变异和 RNA-seq 变异</p>
<h2 id="详细介绍一些算法"><a href="#详细介绍一些算法" class="headerlink" title="详细介绍一些算法"></a>详细介绍一些算法</h2><ul>
<li>1.预处理</li>
</ul>
<p>预处理的主要目的是防止低质量的读取进入不同的评价过程。读取质量通常是通过平均基础质量得分、mapping质量得分和参考基因组的错配次数等来衡量的</p>
<p>gatkindelrealigner 和 BQSR (基准质量得分重新校准)</p>
<p>-提高识别插入的准确性<br>samtools 与 picard </p>
<p>-去除PCR重复(但在基于 pcr 的扩增序列测序应用中，不推荐使用这种方法，因为不同的 DNA 片段可以共享相同的基因组坐标。此步骤中还包括降采样，在此期间随机选择读取的子集以继续进行下一步。如果在特定区域进行降采样，则可以节省计算时间并提高覆盖的一致性，但也会使结果不确定。)</p>
<ul>
<li>2.call 变异</li>
</ul>
<p>各种软件使用各自的算法call 变异</p>
<ul>
<li>3.后期筛选处理</li>
</ul>
<p>测序有些情况会欺骗统计模型将它们作为真正的变异传递。大多数变量调用方应用阈值来识别这些位置，从而提高特异性。例如，链偏置过滤，捕获只有或主要观察到一条链的读数，在 Illumina 公司的一个常见错误读，链偏好过滤器依赖于fisher，以确定不平衡的链分布。许多过滤器集中于重复区域，如homopolymer, microsatellite, or low complexity regions，这些区域已知会由于排列和排序错误而导致错误。硬过滤器用于大多数变异调用者，要么完全拒绝某些区域的变异，要么依赖于经验硬阈值。</p>
<ul>
<li>4.匹配正常肿瘤来call 变异</li>
</ul>
<h2 id="下面介绍一下文章中介绍的各个软件的原理"><a href="#下面介绍一下文章中介绍的各个软件的原理" class="headerlink" title="下面介绍一下文章中介绍的各个软件的原理"></a>下面介绍一下文章中介绍的各个软件的原理</h2><p>其基本思想是利用肿瘤鉴定潜在的变异，并利用匹配的正常样本区分体细胞变异和germline变异丢失(LOH)。<br>VarScan2、 qSNP、 Shimmer、 RADIA、 SOAPsnv 和 VarDict<br>对于一个潜在的 SNV，VarScan2需要至少两个支持读操作和8% 的 VAF (可由用户调整)。其他call 变异软件在他们的算法中有类似的阈值，这些阈值通常设置在一般 NGS 数据的噪声水平之上，并且期望过滤掉低水平的人为影响。接下来，在匹配的正常情况下分析潜在的 SNV 位点，以过滤掉非体细胞变异。2，Shimmer，SOAPsnv，和 VarDict 应用 Fisher 的精确测试22列联表的读数(参考值、非参考值和肿瘤值、正常值)。一个小的 p 值表明非参考读数不成比例地分布在两个样本中，因此提示体细胞变异。qSNP 和 RADIA 应用一系列启发式规则来标记在肿瘤中被充分观察到但在正常情况下弱或未被观察到的体细胞变异体。如果同一患者的 RNA-seq 数据可用，RADIA 将在综合分析中包括基因表达数据，以进一步减少假阳性。</p>
<p>联合基因型分析，SomaticSniper，FaSD-somatic，SAMtools，JointSNVMix2，Virmid，SNVSniffer，Seurat，和 CaVEMan 采用这种方式,假设肿瘤和正常的二倍体，并评估关节的可能性。突变calling是基因推理的必然结果。这些算法的核心是联合基因型的后验概率，由贝叶斯规则计算，即,</p>
<p><img src="/img/calling_mutation/1597125904751.png" alt="1597125904751.png"></p>
<p>其中GT、 GN 为肿瘤基因型，DT 为正常基因型，DN 为肿瘤和正常基因型。先验基因型概率 p (GT，GN)可能依赖于全基因组 SNP 率、体细胞突变率、 Ti-Tv 比值等。数据的联合似然 p (DT，DN | GT，GN)可以通过二项式概率计算，将覆盖一个站点的基地视为独立的伯努利试验，其成功概率取决于基因型和排序错误率。一旦连接基因型被推断出来，体细胞的变异就自然而然地发生了。突变和体细胞突变用“评分”概括了突变的证据，这实质上是肿瘤与正常人具有相同基因型的对数转化概率。其中GT∈{ AA，AC，AG，AT，CC，CG，CT，GG，GT，TT }。体细胞分数较高的部位在肿瘤和正常组织中更可能有不同的基因型，并被鉴定为可能的体细胞变异体。<br>SAMtools 采用了同样的策略，但没有使用后验概率，而是使用对数似然比作为变体评分。<br>JointSNVMix2，Virmid 和 SNVSniffer 将这10个外显基因型分解为 AA，AB 和 BB (a 为参考基因，b 为非参考基因) ，因此将联合基因型分解为33表。体细胞变异调用等价于计算 p (Somatic) = p (AA，AB) + p (AA，BB) ，肿瘤中正常和杂合或纯合非参照的纯合参照概率。具体来说，JointSNVMix2应用一个分层贝叶斯模型来估计联合基因型概率。<br>Virmid 将肿瘤视为正常组织和体细胞突变的混合物，提供了肿瘤中正常组织的联合基因型和比例的联合估计。<br>SNVsniffer 采用启发式和联合基因型分析的混合方法。启发式分析中的高可信体细胞变异被直接报道，低可信变异需要进一步的联合基因型概率估计检验。<br>CaVEMan 应用期望最大化算法估计基因型概率。</p>
<hr>
<p>由于肿瘤样本中存在罕见的异质亚克隆，二倍体假设可能过于简单。为了发现复杂肿瘤基因组中的变异，尤其是罕见亚克隆中的变异体，一些人放弃了二倍体假设和模型联合等位基因频率(fT，fN) ，而代之以联合基因型(GT，GN)。等位基因频率分析方法由斯 MuTect、 LoFreq、 EBCall、 deepSNV、 lololopicker 和 MuSE 使用。</p>
<p>MuTect 将体细胞变异调用定义为两个模型选择问题。在肿瘤方面，对两种模型进行了评价和比较: 野生型模型 M0和突变型模型 Mf，前者假定所有非参考读数均来自技术伪造，后者假定变异等位基因以未知频率存在。通过计算对数似然比(LOD score)来选择更好的拟合模型。潜在突变位点(高 LOD 值) ，另一个模型选择正常进行比较广型模型 M0和杂合子模型 M0.5。如果 M0强烈优于 M0.5，则该变异被标记为躯体变异。LoFreq、 EBCall、 deepSNV 和 lolololopicker 将调用的变量表示为一个假设检验问题，其中零假设是野生型的，替代假设是变异的，检验统计量是观察到的非参考读数 nT。</p>
<p>LoFreq 将每个基地视为一个独立的 Bernoulli 试验，具有不同的“成功”概率，其中成功被定义为非参考，而成功的概率由质量得分决定。在这种情况下，nT 服从泊松-二项分布，p 值可以作为观察到比 nT 更多的非参考读数的概率来计算。</p>
<p>由于已知体细胞变异在某些热点、序列上下文和非编码区域富集，所以 EBCall、 deepSNV 和 LoLoPicker 可以估计特定位点的错误率，因此可以在每个位点设置明确和更准确的检测限。尤其是，deepSNV 和 lololopicker 被设计用于调用具有目标测序数据的低频变异。和 deepSNV 并不依赖于质量分数来推断出错误率，而是假设在每个目标位置，错误率是一个随机变量，遵循一个 Β分布。在无效假设下，nT 遵循一个 β-二项式分布，并相应地计算 p 值。在 EBCall 中，Β分布参数是从其他独立控制样本的序列中获得的。在 deepSNV 中，利用肿瘤样本和正常样本进行参数估计。</p>
<p>LoLoPicker 也采用了类似的策略，但对站点特定的错误率假定为固定值进行了重要修改。对于福尔马林固定和石蜡包埋(FFPE)等低质量样品的变异调用，特定地点的错误率特别有用，因为与新鲜样品相比，这些地点之间的错误率更高，更不平衡。然而，估计特定地点的错误率需要对大量样本进行排序，这并不总是可行的。</p>
<p>MuSE 将体细胞 SNVs 视为 DNA 的结果，并用状态空间为 a，t，g，c 的连续时间马尔可夫过程建模。非参考等位基因的平衡频率比较样本特定的阈值从独立的公共数据集。</p>
<hr>
<p>基于单体类型的策略(相对于主流的基于位置的策略)被结构变异calling软件广泛采用，这些calling 软件需要将读取的内容组合起来以重建长变异。这也是Platypus、 HapMuC、 LocHap、 FreeBayes 和 mutett2使用的检测 SNV 的强大策略。这些算法在局部集合读取一个区域并生成候选单体类型，这些候选单体类型可以用 de Bruijn-like 图表示。每个单体类型的可能性是通过将每个读到的单体类型对齐并计算读支持度来估计的。基于单倍型的变异调用方在变异密集区域具有优势，因为它们不依赖于局部对齐，而局部对齐在困难区域是容易出错的。基于单倍型的调用方还提供了关于变体共存的额外信息。对于基于单体类型的调用方，indel 重新对齐不再有效，因为原始的本地对齐信息被丢弃，读操作被组装和重新对齐。</p>
<hr>
<p>机器学习方法在分类方面发展迅速，而变异调用本质上是一个分类问题。变异 seq，SomaticSeq，snoop 和 BAYSIC 是应用机器学习方法的典型变异调用方。变异 seq 提取每个站点的相关特征，并基于这些特征和一组“基本事实”的体细胞变体训练4个分类器(随机森林、贝叶斯自适应回归树、支持向量机和 Logit模型)。然后在测试数据集上测试训练后的分类器。</p>
<p>SNooPer 训练一个随机森林分类器，用于处理低覆盖率数据。</p>
<p>SomaticSeq 遵循相同的有监督的训练测试过程，但在两个方面不同于 MutationSeq 或 snoop。首先，采用自适应 boosting 算法进行分类。其次，它是一个集成变量调用程序，需要将来自其他软件(MuTect、 SomaticSniper、 VarScan2、 JointSNVMix2和 VarDict)的变量调用联合起来作为起点，然后应用自己的分类器去除假阳性。</p>
<p>BAYSIC 也是一个集合变量调用程序，它使用一个无监督的潜在类模型来组合多个调用。</p>
<div data-largeobj="" data-largeobj-link-rid="largeobj_idm139960369037184" class="xtable"><table frame="hsides" rules="groups" class="rendered small default_table"><thead><tr><th align="left" rowspan="1" colspan="1">Variant caller</th><th align="left" rowspan="1" colspan="1">Type of variant</th><th align="left" rowspan="1" colspan="1">Single-sample mode</th><th align="left" rowspan="1" colspan="1">Type of core algorithm</th></tr></thead><tbody><tr><td align="left" rowspan="1" colspan="1">BAYSIC&nbsp;<a href="/pmc/articles/PMC5852328/#bb0240" rid="bb0240" class=" bibr" target="mainwindow">[48]</a></td><td align="left" rowspan="1" colspan="1">SNV</td><td align="left" rowspan="1" colspan="1">No</td><td align="left" rowspan="1" colspan="1">Machine learning (ensemble caller)</td></tr><tr><td align="left" rowspan="1" colspan="1">CaVEMan&nbsp;<a href="/pmc/articles/PMC5852328/#bb0170" rid="bb0170" class=" bibr" target="mainwindow">[34]</a></td><td align="left" rowspan="1" colspan="1">SNV</td><td align="left" rowspan="1" colspan="1">No</td><td align="left" rowspan="1" colspan="1">Joint genotype analysis</td></tr><tr><td align="left" rowspan="1" colspan="1">deepSNV&nbsp;<a href="/pmc/articles/PMC5852328/#bb0190" rid="bb0190" class=" bibr" target="mainwindow">[38]</a></td><td align="left" rowspan="1" colspan="1">SNV</td><td align="left" rowspan="1" colspan="1">No</td><td align="left" rowspan="1" colspan="1">Allele frequency analysis</td></tr><tr><td align="left" rowspan="1" colspan="1">EBCall&nbsp;<a href="/pmc/articles/PMC5852328/#bb0185" rid="bb0185" class=" bibr" target="mainwindow">[37]</a></td><td align="left" rowspan="1" colspan="1">SNV, indel</td><td align="left" rowspan="1" colspan="1">No</td><td align="left" rowspan="1" colspan="1">Allele frequency analysis</td></tr><tr><td align="left" rowspan="1" colspan="1">FaSD-somatic&nbsp;<a href="/pmc/articles/PMC5852328/#bb0155" rid="bb0155" class=" bibr" target="mainwindow">[31]</a></td><td align="left" rowspan="1" colspan="1">SNV</td><td align="left" rowspan="1" colspan="1">Yes</td><td align="left" rowspan="1" colspan="1">Joint genotype analysis</td></tr><tr><td align="left" rowspan="1" colspan="1">FreeBayes&nbsp;<a href="/pmc/articles/PMC5852328/#bb0220" rid="bb0220" class=" bibr" target="mainwindow">[44]</a></td><td align="left" rowspan="1" colspan="1">SNV, indel</td><td align="left" rowspan="1" colspan="1">Yes</td><td align="left" rowspan="1" colspan="1">Haplotype analysis</td></tr><tr><td align="left" rowspan="1" colspan="1">HapMuC&nbsp;<a href="/pmc/articles/PMC5852328/#bb0210" rid="bb0210" class=" bibr" target="mainwindow">[42]</a></td><td align="left" rowspan="1" colspan="1">SNV, indel</td><td align="left" rowspan="1" colspan="1">Yes</td><td align="left" rowspan="1" colspan="1">Haplotype analysis</td></tr><tr><td align="left" rowspan="1" colspan="1">JointSNVMix2&nbsp;<a href="/pmc/articles/PMC5852328/#bb0150" rid="bb0150" class=" bibr" target="mainwindow">[30]</a></td><td align="left" rowspan="1" colspan="1">SNV</td><td align="left" rowspan="1" colspan="1">No</td><td align="left" rowspan="1" colspan="1">Joint genotype analysis</td></tr><tr><td align="left" rowspan="1" colspan="1">LocHap&nbsp;<a href="/pmc/articles/PMC5852328/#bb0215" rid="bb0215" class=" bibr" target="mainwindow">[43]</a></td><td align="left" rowspan="1" colspan="1">SNV, indel</td><td align="left" rowspan="1" colspan="1">No</td><td align="left" rowspan="1" colspan="1">Haplotype analysis</td></tr><tr><td align="left" rowspan="1" colspan="1">LoFreq&nbsp;<a href="/pmc/articles/PMC5852328/#bb0180" rid="bb0180" class=" bibr" target="mainwindow">[36]</a></td><td align="left" rowspan="1" colspan="1">SNV, indel</td><td align="left" rowspan="1" colspan="1">Yes</td><td align="left" rowspan="1" colspan="1">Allele frequency analysis</td></tr><tr><td align="left" rowspan="1" colspan="1">LoLoPicker&nbsp;<a href="/pmc/articles/PMC5852328/#bb0195" rid="bb0195" class=" bibr" target="mainwindow">[39]</a></td><td align="left" rowspan="1" colspan="1">SNV</td><td align="left" rowspan="1" colspan="1">No</td><td align="left" rowspan="1" colspan="1">Allele frequency analysis</td></tr><tr><td align="left" rowspan="1" colspan="1">MutationSeq&nbsp;<a href="/pmc/articles/PMC5852328/#bb0225" rid="bb0225" class=" bibr" target="mainwindow">[45]</a></td><td align="left" rowspan="1" colspan="1">SNV</td><td align="left" rowspan="1" colspan="1">No</td><td align="left" rowspan="1" colspan="1">Machine learning</td></tr><tr><td align="left" rowspan="1" colspan="1">MuSE&nbsp;<a href="/pmc/articles/PMC5852328/#bb0200" rid="bb0200" class=" bibr" target="mainwindow">[40]</a></td><td align="left" rowspan="1" colspan="1">SNV</td><td align="left" rowspan="1" colspan="1">No</td><td align="left" rowspan="1" colspan="1">Markov chain model</td></tr><tr><td align="left" rowspan="1" colspan="1">MuTect&nbsp;<a href="/pmc/articles/PMC5852328/#bb0175" rid="bb0175" class=" bibr" target="mainwindow">[35]</a></td><td align="left" rowspan="1" colspan="1">SNV</td><td align="left" rowspan="1" colspan="1">Yes</td><td align="left" rowspan="1" colspan="1">Allele frequency analysis</td></tr><tr><td align="left" rowspan="1" colspan="1">SAMtools&nbsp;<a href="/pmc/articles/PMC5852328/#bb0040" rid="bb0040" class=" bibr" target="mainwindow">[8]</a></td><td align="left" rowspan="1" colspan="1">SNV, indel</td><td align="left" rowspan="1" colspan="1">Yes</td><td align="left" rowspan="1" colspan="1">Joint genotype analysis</td></tr><tr><td align="left" rowspan="1" colspan="1">Platypus&nbsp;<a href="/pmc/articles/PMC5852328/#bb0205" rid="bb0205" class=" bibr" target="mainwindow">[41]</a></td><td align="left" rowspan="1" colspan="1">SNV, indel, SV</td><td align="left" rowspan="1" colspan="1">Yes</td><td align="left" rowspan="1" colspan="1">Haplotype analysis</td></tr><tr><td align="left" rowspan="1" colspan="1">qSNP&nbsp;<a href="/pmc/articles/PMC5852328/#bb0120" rid="bb0120" class=" bibr" target="mainwindow">[24]</a></td><td align="left" rowspan="1" colspan="1">SNV</td><td align="left" rowspan="1" colspan="1">No</td><td align="left" rowspan="1" colspan="1">Heuristic threshold</td></tr><tr><td align="left" rowspan="1" colspan="1">RADIA&nbsp;<a href="/pmc/articles/PMC5852328/#bb0130" rid="bb0130" class=" bibr" target="mainwindow">[26]</a></td><td align="left" rowspan="1" colspan="1">SNV</td><td align="left" rowspan="1" colspan="1">No</td><td align="left" rowspan="1" colspan="1">Heuristic threshold</td></tr><tr><td align="left" rowspan="1" colspan="1">Seurat&nbsp;<a href="/pmc/articles/PMC5852328/#bb0165" rid="bb0165" class=" bibr" target="mainwindow">[33]</a></td><td align="left" rowspan="1" colspan="1">SNV, indel, SV</td><td align="left" rowspan="1" colspan="1">No</td><td align="left" rowspan="1" colspan="1">Joint genotype analysis</td></tr><tr><td align="left" rowspan="1" colspan="1">Shimmer&nbsp;<a href="/pmc/articles/PMC5852328/#bb0125" rid="bb0125" class=" bibr" target="mainwindow">[25]</a></td><td align="left" rowspan="1" colspan="1">SNV, indel</td><td align="left" rowspan="1" colspan="1">No</td><td align="left" rowspan="1" colspan="1">Heuristic threshold</td></tr><tr><td align="left" rowspan="1" colspan="1">SNooPer&nbsp;<a href="/pmc/articles/PMC5852328/#bb0235" rid="bb0235" class=" bibr" target="mainwindow">[47]</a></td><td align="left" rowspan="1" colspan="1">SNV, indel</td><td align="left" rowspan="1" colspan="1">Yes</td><td align="left" rowspan="1" colspan="1">Machine learning</td></tr><tr><td align="left" rowspan="1" colspan="1">SNVSniffer&nbsp;<a href="/pmc/articles/PMC5852328/#bb0160" rid="bb0160" class=" bibr" target="mainwindow">[32]</a></td><td align="left" rowspan="1" colspan="1">SNV, indel</td><td align="left" rowspan="1" colspan="1">Yes</td><td align="left" rowspan="1" colspan="1">Joint genotype analysis</td></tr><tr><td align="left" rowspan="1" colspan="1">SOAPsnv&nbsp;<a href="/pmc/articles/PMC5852328/#bb0135" rid="bb0135" class=" bibr" target="mainwindow">[27]</a></td><td align="left" rowspan="1" colspan="1">SNV</td><td align="left" rowspan="1" colspan="1">No</td><td align="left" rowspan="1" colspan="1">Heuristic threshold</td></tr><tr><td align="left" rowspan="1" colspan="1">SomaticSeq&nbsp;<a href="/pmc/articles/PMC5852328/#bb0230" rid="bb0230" class=" bibr" target="mainwindow">[46]</a></td><td align="left" rowspan="1" colspan="1">SNV</td><td align="left" rowspan="1" colspan="1">No</td><td align="left" rowspan="1" colspan="1">Machine learning (ensemble caller)</td></tr><tr><td align="left" rowspan="1" colspan="1">SomaticSniper&nbsp;<a href="/pmc/articles/PMC5852328/#bb0140" rid="bb0140" class=" bibr" target="mainwindow">[28]</a></td><td align="left" rowspan="1" colspan="1">SNV</td><td align="left" rowspan="1" colspan="1">No</td><td align="left" rowspan="1" colspan="1">Joint genotype analysis</td></tr><tr><td align="left" rowspan="1" colspan="1">Strelka&nbsp;<a href="/pmc/articles/PMC5852328/#bb0085" rid="bb0085" class=" bibr" target="mainwindow">[17]</a></td><td align="left" rowspan="1" colspan="1">SNV, indel</td><td align="left" rowspan="1" colspan="1">No</td><td align="left" rowspan="1" colspan="1">Allele frequency analysis</td></tr><tr><td align="left" rowspan="1" colspan="1">TVC&nbsp;<a href="/pmc/articles/PMC5852328/#bb0485" rid="bb0485" class=" bibr" target="mainwindow">[97]</a></td><td align="left" rowspan="1" colspan="1">SNV, indel, SV</td><td align="left" rowspan="1" colspan="1">Yes</td><td align="left" rowspan="1" colspan="1">Ion Torrent specific</td></tr><tr><td align="left" rowspan="1" colspan="1">VarDict&nbsp;<a href="/pmc/articles/PMC5852328/#bb0090" rid="bb0090" class=" bibr" target="mainwindow">[18]</a></td><td align="left" rowspan="1" colspan="1">SNV, indel, SV</td><td align="left" rowspan="1" colspan="1">Yes</td><td align="left" rowspan="1" colspan="1">Heuristic threshold</td></tr><tr><td align="left" rowspan="1" colspan="1">VarScan2&nbsp;<a href="/pmc/articles/PMC5852328/#bb0045" rid="bb0045" class=" bibr" target="mainwindow">[9]</a></td><td align="left" rowspan="1" colspan="1">SNV, indel</td><td align="left" rowspan="1" colspan="1">Yes</td><td align="left" rowspan="1" colspan="1">Heuristic threshold</td></tr><tr><td align="left" rowspan="1" colspan="1">Virmid&nbsp;<a href="/pmc/articles/PMC5852328/#bb0145" rid="bb0145" class=" bibr" target="mainwindow">[29]</a></td><td align="left" rowspan="1" colspan="1">SNV</td><td align="left" rowspan="1" colspan="1">No</td><td align="left" rowspan="1" colspan="1">Joint genotype analysis</td></tr></tbody></table></div></article><div class="post-copyright"><div class="post-copyright__author"><span class="post-copyright-meta">文章作者: </span><span class="post-copyright-info"><a href="https://zhaohongqiangsoliva.github.io">Solivehong</a></span></div><div class="post-copyright__type"><span class="post-copyright-meta">文章链接: </span><span class="post-copyright-info"><a href="https://zhaohongqiangsoliva.github.io/2020/07/14/2020-08-11-%E4%BD%93%E7%BB%86%E8%83%9Ecalling%20%E7%AE%97%E6%B3%95%E7%BB%BC%E8%BF%B0/">https://zhaohongqiangsoliva.github.io/2020/07/14/2020-08-11-%E4%BD%93%E7%BB%86%E8%83%9Ecalling%20%E7%AE%97%E6%B3%95%E7%BB%BC%E8%BF%B0/</a></span></div><div class="post-copyright__notice"><span class="post-copyright-meta">版权声明: </span><span class="post-copyright-info">本博客所有文章除特别声明外，均采用 <a href="https://creativecommons.org/licenses/by-nc-sa/4.0/" target="_blank">CC BY-NC-SA 4.0</a> 许可协议。转载请注明来自 <a href="https://zhaohongqiangsoliva.github.io" target="_blank">Solivehong Blog</a>！</span></div></div><div class="tag_share"><div class="post-meta__tag-list"><a class="post-meta__tags" href="/tags/%E7%94%9F%E4%BF%A1/">生信</a></div><div class="post_share"><div class="social-share" data-image="/img/dna.png" data-sites="facebook,twitter,wechat,weibo,qq"></div><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/butterfly-extsrc/sharejs/dist/css/share.min.css" media="print" onload="this.media='all'"><script src="https://cdn.jsdelivr.net/npm/butterfly-extsrc/sharejs/dist/js/social-share.min.js" defer></script></div></div><nav class="pagination-post" id="pagination"><div class="prev-post pull-left"><a href="/2020/07/14/2020-07-14-%E7%BB%9F%E8%AE%A1%E5%AD%A6%E5%8F%8D%E6%80%9D-00-%E7%9B%AE%E5%BD%95/"><img class="prev-cover" src="/" onerror="onerror=null;src='/img/404.jpg'" alt="cover of previous post"><div class="pagination-info"><div class="label">上一篇</div><div class="prev_info">Statistical Rethinking统计学反思(1章节总结)</div></div></a></div><div class="next-post pull-right"><a href="/2020/07/13/2020-07-13-%E7%94%9F%E4%BF%A1%E5%9F%B9%E8%AE%AD%E6%89%8B%E5%86%8C/"><img class="next-cover" src="/" onerror="onerror=null;src='/img/404.jpg'" alt="cover of next post"><div class="pagination-info"><div class="label">下一篇</div><div class="next_info">生信培训手册</div></div></a></div></nav><div class="relatedPosts"><div class="headline"><i class="fas fa-thumbs-up fa-fw"></i><span>相关推荐</span></div><div class="relatedPosts-list"><div><a href="/2020/06/06/2020-06-06-T%E6%A3%80%E9%AA%8C-%E8%AE%A1%E7%AE%97TMB%E4%B8%A4%E7%A7%8D%E6%96%B9%E5%BC%8F%E5%B7%AE%E5%BC%82/" title="T检验-计算TMB两种方式差异"><img class="cover" src="/" alt="cover"><div class="content is-center"><div class="date"><i class="far fa-calendar-alt fa-fw"></i> 2020-06-06</div><div class="title">T检验-计算TMB两种方式差异</div></div></a></div><div><a href="/2020/06/06/2020-06-06-T%E6%A3%80%E9%AA%8C%E3%80%81F%E6%A3%80%E9%AA%8C%E3%80%81%E5%8D%A1%E6%96%B9%E6%A3%80%E9%AA%8C%E8%AF%A6%E7%BB%86%E5%88%86%E6%9E%90%E5%8F%8A%E5%BA%94%E7%94%A8%E5%9C%BA%E6%99%AF%E6%80%BB%E7%BB%93/" title="T检验、F检验、卡方检验详细分析及应用场景总结"><img class="cover" src="/" alt="cover"><div class="content is-center"><div class="date"><i class="far fa-calendar-alt fa-fw"></i> 2020-06-06</div><div class="title">T检验、F检验、卡方检验详细分析及应用场景总结</div></div></a></div><div><a href="/2020/06/06/2020-06-06-%E9%A9%B1%E5%8A%A8%E5%9F%BA%E5%9B%A0%E6%A3%80%E6%B5%8B/" title="驱动基因"><img class="cover" src="/" alt="cover"><div class="content is-center"><div class="date"><i class="far fa-calendar-alt fa-fw"></i> 2020-06-06</div><div class="title">驱动基因</div></div></a></div><div><a href="/2020/07/13/2020-07-13-%E7%94%9F%E4%BF%A1%E5%9F%B9%E8%AE%AD%E6%89%8B%E5%86%8C/" title="生信培训手册"><img class="cover" src="/" alt="cover"><div class="content is-center"><div class="date"><i class="far fa-calendar-alt fa-fw"></i> 2020-07-13</div><div class="title">生信培训手册</div></div></a></div><div><a href="/2020/06/09/2020-06-09-CNV-Facets%E8%BD%AF%E4%BB%B6%E6%A3%80%E6%B5%8Bcnv/" title="FACETS Testing CNV"><img class="cover" src="/" alt="cover"><div class="content is-center"><div class="date"><i class="far fa-calendar-alt fa-fw"></i> 2020-06-09</div><div class="title">FACETS Testing CNV</div></div></a></div></div></div></div><div class="aside-content" id="aside-content"><div class="card-widget card-info"><div class="is-center"><div class="avatar-img"><img src="/img/dna.png" onerror="this.onerror=null;this.src='/img/friend_404.gif'" alt="avatar"/></div><div class="author-info__name">Solivehong</div><div class="author-info__description"></div></div><div class="card-info-data site-data is-center"><a href="/archives/"><div class="headline">文章</div><div class="length-num">19</div></a><a href="/tags/"><div class="headline">标签</div><div class="length-num">6</div></a><a href="/categories/"><div class="headline">分类</div><div class="length-num">0</div></a></div><a id="card-info-btn" target="_blank" rel="noopener" href="https://github.com/xxxxxx"><i class="fab fa-github"></i><span>Follow Me</span></a><div class="card-info-social-icons is-center"><a class="social-icon" href="https://github.com/zhaohongqiangsoliva" target="_blank" title="Github"><i class="fab fa-github"></i></a><a class="social-icon" href="mailto:zhaohongqiangsoliva@gmail.com" target="_blank" title="Email"><i class="fas fa-envelope"></i></a></div></div><div class="card-widget card-announcement"><div class="item-headline"><i class="fas fa-bullhorn fa-shake"></i><span>公告</span></div><div class="announcement_content">This is my Blog</div></div><div class="sticky_layout"><div class="card-widget" id="card-toc"><div class="item-headline"><i class="fas fa-stream"></i><span>目录</span><span class="toc-percentage"></span></div><div class="toc-content"><ol class="toc"><li class="toc-item toc-level-1"><a class="toc-link" href="#%E4%BD%93%E7%BB%86%E8%83%9Ecalling-%E7%AE%97%E6%B3%95%E7%BB%BC%E8%BF%B0-somatic-single-nucleotide-variant-calling-algorithms"><span class="toc-number">1.</span> <span class="toc-text">体细胞calling 算法综述 somatic single nucleotide variant calling algorithms</span></a><ol class="toc-child"><li class="toc-item toc-level-2"><a class="toc-link" href="#paper%E5%9F%BA%E6%9C%AC%E4%BF%A1%E6%81%AF"><span class="toc-number">1.1.</span> <span class="toc-text">paper基本信息</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#%E6%80%BB%E7%BB%93"><span class="toc-number">1.2.</span> <span class="toc-text">总结</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#%E8%AF%A6%E7%BB%86%E4%BB%8B%E7%BB%8D%E4%B8%80%E4%BA%9B%E7%AE%97%E6%B3%95"><span class="toc-number">1.3.</span> <span class="toc-text">详细介绍一些算法</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#%E4%B8%8B%E9%9D%A2%E4%BB%8B%E7%BB%8D%E4%B8%80%E4%B8%8B%E6%96%87%E7%AB%A0%E4%B8%AD%E4%BB%8B%E7%BB%8D%E7%9A%84%E5%90%84%E4%B8%AA%E8%BD%AF%E4%BB%B6%E7%9A%84%E5%8E%9F%E7%90%86"><span class="toc-number">1.4.</span> <span class="toc-text">下面介绍一下文章中介绍的各个软件的原理</span></a></li></ol></li></ol></div></div><div class="card-widget card-recent-post"><div class="item-headline"><i class="fas fa-history"></i><span>最新文章</span></div><div class="aside-list"><div class="aside-list-item"><a class="thumbnail" href="/2022/10/27/2022-10-27-fzf%E7%9A%84%E4%BD%BF%E7%94%A8/" title="Fzf的使用"><img src="https://unpkg.com/justlovesmile-img/cover5.JPG" onerror="this.onerror=null;this.src='/img/404.jpg'" alt="Fzf的使用"/></a><div class="content"><a class="title" href="/2022/10/27/2022-10-27-fzf%E7%9A%84%E4%BD%BF%E7%94%A8/" title="Fzf的使用">Fzf的使用</a><time datetime="2022-10-26T16:00:00.000Z" title="发表于 2022-10-27 00:00:00">2022-10-27</time></div></div><div class="aside-list-item"><a class="thumbnail" href="/2022/10/27/2022-10-27-linuxTools%E6%89%8B%E5%86%8C%E5%A4%A7%E5%85%A8/" title="linux 命令大全"><img src="https://unpkg.com/justlovesmile-img/cover3.jpg" onerror="this.onerror=null;this.src='/img/404.jpg'" alt="linux 命令大全"/></a><div class="content"><a class="title" href="/2022/10/27/2022-10-27-linuxTools%E6%89%8B%E5%86%8C%E5%A4%A7%E5%85%A8/" title="linux 命令大全">linux 命令大全</a><time datetime="2022-10-26T16:00:00.000Z" title="发表于 2022-10-27 00:00:00">2022-10-27</time></div></div><div class="aside-list-item"><a class="thumbnail" href="/2021/10/02/2021-10-02-%E5%BC%BA%E5%8C%96%E5%AD%A6%E4%B9%A0-02-DP/" title="强化学习-动态规划-DP"><img src="https://unpkg.com/justlovesmile-img/cover2.jpg" onerror="this.onerror=null;this.src='/img/404.jpg'" alt="强化学习-动态规划-DP"/></a><div class="content"><a class="title" href="/2021/10/02/2021-10-02-%E5%BC%BA%E5%8C%96%E5%AD%A6%E4%B9%A0-02-DP/" title="强化学习-动态规划-DP">强化学习-动态规划-DP</a><time datetime="2021-10-01T16:00:00.000Z" title="发表于 2021-10-02 00:00:00">2021-10-02</time></div></div><div class="aside-list-item"><a class="thumbnail" href="/2021/10/02/2021-10-02-%E5%BC%BA%E5%8C%96%E5%AD%A6%E4%B9%A0-03-MCMC/" title="蒙特卡洛（Markov Chain &amp; Monte Carlo, MCMC）方法"><img src="https://unpkg.com/justlovesmile-img/cover1.JPG" onerror="this.onerror=null;this.src='/img/404.jpg'" alt="蒙特卡洛（Markov Chain &amp; Monte Carlo, MCMC）方法"/></a><div class="content"><a class="title" href="/2021/10/02/2021-10-02-%E5%BC%BA%E5%8C%96%E5%AD%A6%E4%B9%A0-03-MCMC/" title="蒙特卡洛（Markov Chain &amp; Monte Carlo, MCMC）方法">蒙特卡洛（Markov Chain &amp; Monte Carlo, MCMC）方法</a><time datetime="2021-10-01T16:00:00.000Z" title="发表于 2021-10-02 00:00:00">2021-10-02</time></div></div><div class="aside-list-item"><a class="thumbnail" href="/2021/10/02/2021-10-02-%E5%BC%BA%E5%8C%96%E5%AD%A6%E4%B9%A0-04-%E5%85%8D%E6%A8%A1%E5%9E%8B%E5%AD%A6%E4%B9%A0%E9%A2%84%E6%B5%8B/" title="强化学习—免模型预测"><img src="https://unpkg.com/justlovesmile-img/cover5.JPG" onerror="this.onerror=null;this.src='/img/404.jpg'" alt="强化学习—免模型预测"/></a><div class="content"><a class="title" href="/2021/10/02/2021-10-02-%E5%BC%BA%E5%8C%96%E5%AD%A6%E4%B9%A0-04-%E5%85%8D%E6%A8%A1%E5%9E%8B%E5%AD%A6%E4%B9%A0%E9%A2%84%E6%B5%8B/" title="强化学习—免模型预测">强化学习—免模型预测</a><time datetime="2021-10-01T16:00:00.000Z" title="发表于 2021-10-02 00:00:00">2021-10-02</time></div></div></div></div></div></div></main><footer id="footer"><div id="footer-wrap"><div class="copyright">&copy;2020 - 2023 By Solivehong</div><div class="framework-info"><span>框架 </span><a target="_blank" rel="noopener" href="https://hexo.io">Hexo</a><span class="footer-separator">|</span><span>主题 </span><a target="_blank" rel="noopener" href="https://github.com/jerryc127/hexo-theme-butterfly">Butterfly</a></div></div></footer></div><div id="rightside"><div id="rightside-config-hide"><button id="readmode" type="button" title="阅读模式"><i class="fas fa-book-open"></i></button><button id="darkmode" type="button" title="浅色和深色模式转换"><i class="fas fa-adjust"></i></button><button id="hide-aside-btn" type="button" title="单栏和双栏切换"><i class="fas fa-arrows-alt-h"></i></button></div><div id="rightside-config-show"><button id="rightside_config" type="button" title="设置"><i class="fas fa-cog fa-spin"></i></button><button class="close" id="mobile-toc-button" type="button" title="目录"><i class="fas fa-list-ul"></i></button><button id="go-up" type="button" title="回到顶部"><i class="fas fa-arrow-up"></i></button></div></div><div id="local-search"><div class="search-dialog"><nav class="search-nav"><span class="search-dialog-title">搜索</span><span id="loading-status"></span><button class="search-close-button"><i class="fas fa-times"></i></button></nav><div class="is-center" id="loading-database"><i class="fas fa-spinner fa-pulse"></i><span>  数据库加载中</span></div><div class="search-wrap"><div id="local-search-input"><div class="local-search-box"><input class="local-search-box--input" placeholder="搜索文章" type="text"/></div></div><hr/><div id="local-search-results"></div></div></div><div id="search-mask"></div></div><div><script src="/js/utils.js"></script><script src="/js/main.js"></script><script src="https://cdn.jsdelivr.net/npm/@fancyapps/ui/dist/fancybox.umd.min.js"></script><script src="/js/search/local-search.js"></script><div class="js-pjax"><script>if (!window.MathJax) {
  window.MathJax = {
    tex: {
      inlineMath: [ ['$','$'], ["\\(","\\)"]],
      tags: 'ams'
    },
    chtml: {
      scale: 1.2
    },
    options: {
      renderActions: {
        findScript: [10, doc => {
          for (const node of document.querySelectorAll('script[type^="math/tex"]')) {
            const display = !!node.type.match(/; *mode=display/)
            const math = new doc.options.MathItem(node.textContent, doc.inputJax[0], display)
            const text = document.createTextNode('')
            node.parentNode.replaceChild(text, node)
            math.start = {node: text, delim: '', n: 0}
            math.end = {node: text, delim: '', n: 0}
            doc.math.push(math)
          }
        }, ''],
        insertScript: [200, () => {
          document.querySelectorAll('mjx-container:not\([display]\)').forEach(node => {
            const target = node.parentNode
            if (target.nodeName.toLowerCase() === 'li') {
              target.parentNode.classList.add('has-jax')
            } else {
              target.classList.add('has-jax')
            }
          });
        }, '', false]
      }
    }
  }
  
  const script = document.createElement('script')
  script.src = 'https://cdn.jsdelivr.net/npm/mathjax/es5/tex-mml-chtml.min.js'
  script.id = 'MathJax-script'
  script.async = true
  document.head.appendChild(script)
} else {
  MathJax.startup.document.state(0)
  MathJax.texReset()
  MathJax.typeset()
}</script></div><script async data-pjax src="//busuanzi.ibruce.info/busuanzi/2.3/busuanzi.pure.mini.js"></script></div></body></html>